pkg/apis/core/validation/validation.go

```diff
@@ -1526,247 +1530,287 @@ var supportedDataSourceAPIGroupKinds = map[schema.GroupKind]bool{
+		}
+	}
+
+	if len(pvSpec.AccessModes) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("accessModes"), ""))
 	}
-	for _, mode := range pv.Spec.AccessModes {
+	for _, mode := range pvSpec.AccessModes {
 		if !supportedAccessModes.Has(string(mode)) {
-			allErrs = append(allErrs, field.NotSupported(specPath.Child("accessModes"), mode, supportedAccessModes.List()))
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("accessModes"), mode, supportedAccessModes.List()))
 		}
 	}
 
-	if len(pv.Spec.Capacity) == 0 {
-		allErrs = append(allErrs, field.Required(specPath.Child("capacity"), ""))
-	}
+	if !validateInlinePersistentVolume {
+		if len(pvSpec.Capacity) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("capacity"), ""))
+		}
 
-	if _, ok := pv.Spec.Capacity[core.ResourceStorage]; !ok || len(pv.Spec.Capacity) > 1 {
-		allErrs = append(allErrs, field.NotSupported(specPath.Child("capacity"), pv.Spec.Capacity, []string{string(core.ResourceStorage)}))
-	}
-	capPath := specPath.Child("capacity")
-	for r, qty := range pv.Spec.Capacity {
-		allErrs = append(allErrs, validateBasicResource(qty, capPath.Key(string(r)))...)
-		allErrs = append(allErrs, ValidatePositiveQuantityValue(qty, capPath.Key(string(r)))...)
-	}
-	if len(string(pv.Spec.PersistentVolumeReclaimPolicy)) > 0 {
-		if !supportedReclaimPolicy.Has(string(pv.Spec.PersistentVolumeReclaimPolicy)) {
-			allErrs = append(allErrs, field.NotSupported(specPath.Child("persistentVolumeReclaimPolicy"), pv.Spec.PersistentVolumeReclaimPolicy, supportedReclaimPolicy.List()))
+		if _, ok := pvSpec.Capacity[core.ResourceStorage]; !ok || len(pvSpec.Capacity) > 1 {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("capacity"), pvSpec.Capacity, []string{string(core.ResourceStorage)}))
+		}
+		capPath := fldPath.Child("capacity")
+		for r, qty := range pvSpec.Capacity {
+			allErrs = append(allErrs, validateBasicResource(qty, capPath.Key(string(r)))...)
+			allErrs = append(allErrs, ValidatePositiveQuantityValue(qty, capPath.Key(string(r)))...)
 		}
 	}
 
-	nodeAffinitySpecified, errs := validateVolumeNodeAffinity(pv.Spec.NodeAffinity, specPath.Child("nodeAffinity"))
-	allErrs = append(allErrs, errs...)
+	if len(string(pvSpec.PersistentVolumeReclaimPolicy)) > 0 {
+		if validateInlinePersistentVolume {
+			if pvSpec.PersistentVolumeReclaimPolicy != core.PersistentVolumeReclaimRetain {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("persistentVolumeReclaimPolicy"), "may only be "+string(core.PersistentVolumeReclaimRetain)+" in the context of inline volumes"))
+			}
+		} else {
+			if !supportedReclaimPolicy.Has(string(pvSpec.PersistentVolumeReclaimPolicy)) {
+				allErrs = append(allErrs, field.NotSupported(fldPath.Child("persistentVolumeReclaimPolicy"), pvSpec.PersistentVolumeReclaimPolicy, supportedReclaimPolicy.List()))
+			}
+		}
+	}
 
+	var nodeAffinitySpecified bool
+	var errs field.ErrorList
+	if pvSpec.NodeAffinity != nil {
```

Comment:
same here I think.

Above we have
```
if validateInlinePersistentVolume:
  validate field A
  validate field B
if !validateInlinePersistentVolume:
  validate field A
```

then down here it looks like we've changed to
```
validate field A:
  if validateInlinePersistentVolume:
    validate this way
  if !validateInlinePersistentVolume
    validate that way
```

could we just refactor this all into the above way? Maybe I'm just confused. If it makes more sense to do it this way then please keep it

Reply:
I agree that mixing the two is not ideal. I wanted to keep the validations for existing fields in the same order as it is today by inlining the checks for validateInlinePersistentVolume but could not fit the new non-null/empty checks for validateInlinePersistentVolume in the same pattern. Will try to refactor this a bit.